frontend/src/metabase/query_builder/components/chart-type-selector/use-chart-type-visualizations.ts

```diff
@@ -0,0 +1,95 @@
+import { useCallback, useMemo } from "react";
+import _ from "underscore";
+
+import visualizations from "metabase/visualizations";
+import { sanatizeResultData } from "metabase/visualizations/shared/utils/data";
+import type Question from "metabase-lib/v1/Question";
+import {
+  type CardDisplayType,
+  type Dataset,
+  type VisualizationDisplay,
+  isCardDisplayType,
+} from "metabase-types/api";
+
+import { DEFAULT_VIZ_ORDER } from "./viz-order";
+
+export type UseChartTypeVisualizationsProps = {
+  question: Question;
+  onUpdateQuestion: (question: Question) => void;
+};
+
+export const useChartTypeVisualizations = ({
+  question,
+  onUpdateQuestion,
+  result,
+}: UseChartTypeVisualizationsProps & GetSensibleVisualizationsProps) => {
+  const selectedVisualization = question.display();
+
+  const updateQuestionVisualization = useCallback(
+    (display: CardDisplayType) => {
+      let newQuestion = question.setDisplay(display).lockDisplay(); // prevent viz auto-selection
+      const visualization = visualizations.get(display);
+      if (visualization?.onDisplayUpdate) {
+        const updatedSettings = visualization.onDisplayUpdate(
+          newQuestion.settings(),
+        );
+        newQuestion = newQuestion.setSettings(updatedSettings);
+      }
+
+      onUpdateQuestion(newQuestion);
+    },
+    [onUpdateQuestion, question],
+  );
+
+  const { sensibleVisualizations, nonSensibleVisualizations } = useMemo(
+    () => getSensibleVisualizations({ result }),
+    [result],
+  );
+
+  return {
+    selectedVisualization,
+    updateQuestionVisualization,
+    sensibleVisualizations,
+    nonSensibleVisualizations,
+  };
+};
+
+type IsSensibleVisualizationProps = {
+  result: Dataset | null;
+  vizType: VisualizationDisplay;
+};
+
+const isSensibleVisualization = ({
+  result,
+  vizType,
+}: IsSensibleVisualizationProps) => {
+  const visualization = visualizations.get(vizType);
+  return (
+    (result?.data &&
+      visualization?.isSensible?.(sanatizeResultData(result.data))) ||
+    false
+  );
+};
+
+export type GetSensibleVisualizationsProps = Pick<
+  IsSensibleVisualizationProps,
```

Comment:
nit: I found it a bit harder to follow prop types definitions when they refer to each other like this using `Pick`. It seems that 1-2 props could've been defined in their own own interface without referring to other props